examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp

```diff
@@ -0,0 +1,679 @@
+    user.lastModifiedBy     = userInDb.lastModifiedBy;
+
+    ChipLogDetail(Zcl,
+                  "Found occupied user "
+                  "[endpoint=%d,adjustedIndex=%hu,name=\"%.*s\",credentialsCount=%u,uniqueId=%x,type=%u,credentialRule=%u,"
+                  "createdBy=%d,lastModifiedBy=%d]",
+                  mEndpointId, adjustedUserIndex, static_cast<int>(user.userName.size()), user.userName.data(),
+                  static_cast<unsigned int>(user.credentials.size()), user.userUniqueId, to_underlying(user.userType),
+                  to_underlying(user.credentialRule), user.createdBy, user.lastModifiedBy);
+
+    return true;
+}
+
+bool LockEndpoint::SetUser(uint16_t userIndex, chip::FabricIndex creator, chip::FabricIndex modifier,
+                           const chip::CharSpan & userName, uint32_t uniqueId, UserStatusEnum userStatus, UserTypeEnum usertype,
+                           CredentialRuleEnum credentialRule, const CredentialStruct * credentials, size_t totalCredentials)
+{
+    ChipLogProgress(Zcl,
+                    "Lock App: LockEndpoint::SetUser "
+                    "[endpoint=%d,userIndex=%u,creator=%d,modifier=%d,userName=\"%.*s\",uniqueId=%" PRIx32
+                    ",userStatus=%u,userType=%u,"
+                    "credentialRule=%u,credentials=%p,totalCredentials=%u]",
+                    mEndpointId, userIndex, creator, modifier, static_cast<int>(userName.size()), userName.data(), uniqueId,
+                    to_underlying(userStatus), to_underlying(usertype), to_underlying(credentialRule), credentials,
+                    static_cast<unsigned int>(totalCredentials));
+
+    auto adjustedUserIndex = static_cast<uint16_t>(userIndex - 1);
+    if (adjustedUserIndex > mLockUsers.size())
+    {
+        ChipLogError(Zcl, "Cannot set user - index out of range [endpoint=%d,index=%d,adjustedUserIndex=%u]", mEndpointId,
+                     userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    auto & userInStorage = mLockUsers[adjustedUserIndex];
+    if (userName.size() > DOOR_LOCK_MAX_USER_NAME_SIZE)
+    {
+        ChipLogError(Zcl, "Cannot set user - user name is too long [endpoint=%d,index=%d,adjustedUserIndex=%u]", mEndpointId,
+                     userIndex, adjustedUserIndex);
+        return false;
+    }
+
+    if (totalCredentials > userInStorage.credentials.capacity())
+    {
+        ChipLogError(Zcl,
+                     "Cannot set user - total number of credentials is too big [endpoint=%d,index=%d,adjustedUserIndex=%u"
+                     ",totalCredentials=%u,maxNumberOfCredentials=%u]",
+                     mEndpointId, userIndex, adjustedUserIndex, static_cast<unsigned int>(totalCredentials),
+                     static_cast<unsigned int>(userInStorage.credentials.capacity()));
+        return false;
+    }
+
+    chip::Platform::CopyString(userInStorage.userName, userName);
+    userInStorage.userName[userName.size()] = 0;
```

Comment:
All of the zero-terminating logic is error-prone. Please use std::string for the `userName` to ensure no issues with null-termination. This is your code so you are allowed to do as you wish.

Reply:
Actually, the code is using "flat" storage of username :( That means it expects null-termination in storage as well. This is error-prone and dangerous, but would require significant code changes. A better alternative is to keep username size in the structure, separate from the char array, and avoid any use of `strlen()`, and always generate the `CharSpan`s from explicit size kept alongside the string.

Reply:
Fixed with CharSpan done. PTAL